### PR TITLE
[WIP] [Analysis] File Metadata Backend [PLAT-819]

### PR DESCRIPTION
**Pull request by caseyrollins**
#### Purpose
Analysis and early implementation of the File Metadata backend and internal API.
Note that this PR probably barely functions if checked out and is only meant to demonstrate approach. 

#### Changes
* Generalizes the `MetaSchema` model.
```
class AbstractMetaSchema(ObjectIDMixin, BaseModel):
    ...

class RegistrationMetaSchema(AbstractMetaSchema):
    ...

class FileMetadataMetaSchema(AbstractMetaSchema):
    ...
```

* Adds `FileMetadataRecord` model.
```
class FileMetadataRecord(ObjectIDMixin, BaseModel):

    metadata = DateTimeAwareJSONField(default=dict, blank=True)
    file = models.ForeignKey(OsfStorageFile, related_name='records')
    schema = models.ForeignKey(FileMetadataMetaSchema, related_name='records')

    class Meta:
        unique_together = ('file', 'schema')
```

* Adds base `FileMetadataFormatter` class.
  * Formatter classes control how a given metadata record will be serialized. 
  * For every new file metadata format we support, a corresponding subclass will need to be added.

* Adds `DataCiteMetadataFormatter(FileMetadataFormatter)` subclass.
  * Implements JSON serialization of DataCite file metadata.
  * General DataCite JSON formatting helpers added to metadata.py.

* Adds WIP FileMetadataSerializer to the API.

#### Internal API
```
# Format a file's metadata as JSON
osfs_file.records.get(schema___id='datacite').format()

osfs_file.records.get(schema___id='datacite').format(type='json')

for metadata_record in osfs_file.records.all():
    metadata_record.format()

# Format a file's metadata as XML
osfs_file.records.get(schema___id='datacite').format(type='xml')
```

#### TODOs
- [ ] Write migration to add `FileMetadataMetaSchemas` to the database (i.e. datacite.json). 
- [ ] Write migration to add a `FileMetadataRecord(schema='datacite')` to all existing OSFS files.
- [ ] Add validation for metadata updates. 
- [ ] Add node logging when metadata is updated.
- [ ] Copy file metadata to forked files.
- [ ] Soooo many tests.
- [ ] Probably a lot of other things.

#### Ticket
* [PLAT-819](https://openscience.atlassian.net/browse/PLAT-819)

### CODE REVIEW

File: website/project/metadata/datacite.json [hunk 1]
```diff
@@ -0,0 +1,111 @@
+{
+    "_id": "datacite",
+    "name": "Datacite",
+    "schema_version": 1,
+    "category": "file",
+    "schema": {
```
Comment (sloria) [inline website/project/metadata/datacite.json]: Notes

* Look into JSON schema
* Validate metadata records against this schema

See registration metaschemas (validate_metadata) for examples.

**Comment by sloria** [comment]
Notes:

* Create metadata records for each metaschema when a file is created (see `post_save` signals).

File: website/project/metadata/datacite.json (hunk 1 above)
Comment (sloria) [reply website/project/metadata/datacite.json]: You have to do some finagling with `not` to handle the non-editable fields:

```python
from jsonschema import validate

# A sample schema, like what we'd get from json.load()
schema = {
    "type": "object",
    "properties" : {
         # NOTE: readOnly is only an annotation--it isn't used for validation
        "title": {"type" : "number", 'readOnly': True},
        "description" : {"type": "string", 'readOnly': True},
        "funder" : {"type" : "string"},
    },
    'not': {
        'anyOf': [
            {'required': ['title']},
            {'required': ['description']}
        ]
    }
}

validate({'funder': 'ljaf'}, schema)  # passes

validate({'funder': 'ljaf', 'title': 'nope'}, schema)  # fails
# jsonschema.exceptions.ValidationError: {'anyOf': [{'required': ['title']}, {'required': ['description']}]} is not allowed for {'funder': 'ljaf', 'title': 'nope'}
#
# Failed validating 'not' in schema:
#     {'not': {'anyOf': [{'required': ['title']},
#                        {'required': ['description']}]},
#      'properties': {'description': {'readOnly': True, 'type': 'string'},
#                     'funder': {'type': 'string'},
#                     'title': {'readOnly': True, 'type': 'number'}},
#      'type': 'object'}
#
# On instance:
#     {'funder': 'ljaf', 'title': 'nope'}
```

File: api/metadata/formatters.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    datacite_format_subjects_json,
+)
+
+class FileMetadataFormatter(object):
+
+    @property
+    def metaschema_id(self):
+        raise NotImplementedError('FileMetadataFormatter subclasses must define metaschema_id.')
+
+    def validate(self, metadata):
```
Comment (caseyrollins) [inline api/metadata/formatters.py]: This will be implemented here (on the base class) because `validate()` should just called jsonschema's `validate()`

**Comment by caseyrollins** [comment]
```
osf/
  utils/
    metadata/
      formatters.py
      utils.py
      schemas/
        datacite.json
```

**Comment by sloria** [comment]
In-person pass finished .

**Comment by caseyrollins** [comment]
Closing since this was just an analysis PR.

File: osf/models/metadata.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    metadata = DateTimeAwareJSONField(default=dict, blank=True)
+
+    file = models.ForeignKey(OsfStorageFile, related_name='records', on_delete=models.SET_NULL, null=True)
+    schema = models.ForeignKey(FileMetadataMetaSchema, related_name='records', on_delete=models.SET_NULL, null=True)
+
+    class Meta:
+        unique_together = ('file', 'schema')
+
+    def format(self, type='json'):
+        return self.schema.formatter.format(self.file, self.metadata, type)
```
Comment (caseyrollins) [inline osf/models/metadata.py]: Remove self.file and self.metadata args here. Just type

Comment (caseyrollins) [reply osf/models/metadata.py]: Naming needs work. Potentially `self.schema.serializer.serialize()` (defaults to json) or `self.schema.serializer.serialize(format='xml')`

File: osf/models/metaschema.py [hunk 4]
```diff
@@ -21,8 +24,25 @@ class MetaSchema(ObjectIDMixin, BaseModel):
...
 
     class Meta:
+        abstract = True
         unique_together = ('name', 'schema_version')
 
+
+class FileMetadataMetaSchema(AbstractMetaSchema):
+
+    @property
+    def formatter(self):
```
Comment (caseyrollins) [inline osf/models/metaschema.py]: Replace this property with a `serializer_class` attribute

**Pull request closed without merging**